app/forms/questionnaire_form.py

```diff
@@ -1,26 +1,74 @@
 import logging
 
+from wtforms import validators
+
 from flask_wtf import FlaskForm
 
-from app.forms.date_form import get_date_data, get_date_range_fields
 from app.forms.fields import get_field
 from app.helpers.schema_helper import SchemaHelper
+from app.validation.validators import DateRangeCheck
 
 from werkzeug.datastructures import MultiDict
 
 logger = logging.getLogger(__name__)
 
 
-def get_date_range_field(question, data, error_messages):
-    to_field_id = question['answers'][1]['id']
-    to_field_data = get_date_data(data, to_field_id)
+class QuestionnaireForm(FlaskForm):
+
+    def __init__(self, block_json, formdata=None, **kwargs):
+        self.block_json = block_json
+        self.question_errors = {}
+
+        if formdata:
+            super(QuestionnaireForm, self).__init__(formdata=formdata, **kwargs)
+        else:
+            super(QuestionnaireForm, self).__init__(**kwargs)
+
+    def validate(self):
+        """
+        Validate this form as usual and check for any form-level date-range validation errors
+        :return:
+        """
+        valid_form = True
+        valid_fields = FlaskForm.validate(self)
+
+        for question_id, period_from_id, period_to_id in self.date_ranges:
+            if period_from_id not in self.errors and period_to_id not in self.errors:
+                period_from = getattr(self, period_from_id)
+                period_to = getattr(self, period_to_id)
+                validator = DateRangeCheck()
+
+                # Check every field on each form has populated data
+                populated = all([f.data for f in period_from] + [f.data for f in period_to])
+
+                if populated:
+                    try:
+                        validator(self, period_from, period_to)
+                    except validators.ValidationError as e:
+                        self.question_errors[question_id] = str(e)
+                        valid_form = False
```

Comment:
We could just have valid_fields and just set that to false. I am not too sure what the difference between them is.